snuba/clickhouse/processors.py

```diff
@@ -0,0 +1,25 @@
+from abc import ABC, abstractmethod
+
+from snuba.clickhouse.query import Query
+from snuba.request.request_settings import RequestSettings
+
+
+class QueryProcessor(ABC):
+    """
+    A transformation applied to a Clickhouse Query. This transformation mutates the
+    Query object in place.
+
+    Processors that extend this class are executed during the Clickhouse specific part
+    of the query execution pipeline.
+    As their logical counterparts, Clickhouse query processors are stateless and are
+    independent from each other. Each processor must leave the query in a valid state
+    and must not depend on the execution of another processor before or after.
+
+    Processors are designed to be stateless. There is no guarantee whether the same
+    instance may be reused.
+    """
+
+    @abstractmethod
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        # TODO: Make the Query class immutable.
```

Comment:
Contrary to the docstring: well yes, the docstring reflects the current state, this is a TODO. When/if we do this the docstring will change accordingly.

I think immutability would still be useful, specifically with the AST (expressions are immutable already, so replacing conditions will work differently than today).
There are still some cases where I think side effects are dangerous, they materialize in tracing now, (we store the query in a span, then we mutate it and when the span is recorded we record the final value instead of the initial one). This is a relatively small cases but it shows the danger. 

I think we should revisit this point when we get there. I will turn the TODO into "consider making the query immutable"

Reply:
Alright, thanks for clarifying. To be clear: I wasn't necessarily _advocating_ for making the query mutable, just noting it doesn't seem to be as urgent of an issue as I thought it was previously.

> expressions are immutable already, so replacing conditions will work differently than today

Ah, good point.